src/core/pattern.js

```diff
@@ -167,10 +167,9 @@ class RadialAxialShading extends BaseShading {
     const fnObj = dict.getRaw("Function");
     const fn = pdfFunctionFactory.createFromArray(fnObj);
 
-    // 10 samples seems good enough for now, but probably won't work
-    // if there are sharp color changes. Ideally, we would implement
-    // the spec faithfully and add lossless optimizations.
-    const NUMBER_OF_SAMPLES = 10;
+    // Use lcm(1,2,3,4,5,6,7,8,10) = 840 (including 9 increases this to 2520)
+    // to catch evenly spaced stops. oeis.org/A003418
+    const NUMBER_OF_SAMPLES = 840;
```

Comment:
So this evaluates the function quite a bit, but in the end most gradients (the ones that already existed in the test corpus) now only have 2 color stops instead of 10. Obviously, if the application requires more then more will be used. For example 10572 uses 48 (a lot is clipped and not shown) and 14165 uses 113 (but that's about right for approximating nine sine waves at ~10 each).

I'll take a look into the benchmarking.

Reply:
I have attempted local benchmarking with a few gradient heavy pdfs, including issue16127.pdf (thanks for fixing that issue to make this possible) since it contains a very large number (100s) of axial and radial (Type 2 and 3) shaders with Stitching with Exponential Interpolation (Type 3 with Type 2) functions. (Though should test with  PostScript (Type 4) functions which should be the worst case for this change.) However, all the timings I'm getting locally are mostly (almost entirely) noise. At the very least this change doesn't appear to rise above the noise level, but the noise level is quite high.

To try to get some handle on real timing, I opened issue16127.pdf  with pdf.js with this change and used the Firefox Performance Profiler to grab a trace from loading to the first page displayed. Not surprisingly this spends ~25% of its time in `showType3Text` since that's the majority of the drawing. In this profile `showType3Text`  has a cumulative 2764 samples with with `RadialAxialShadingPattern` contributing 5 of those samples. (Interestingly, it appears that the cumulative time in `CanvasGradient.addColorStop` over on the display side consistently takes longer. That interface looks like it could benefit from a bulk add.)

I'll see about trying to find or create a pdf with many axial or radial shaders that use PostScript functions and see if that is measurable.

It looks like [pdfium uses 256 samples](https://pdfium.googlesource.com/pdfium/+/e78b735937cebb45183606c538c09c5198524b5d/core/fpdfapi/render/cpdf_rendershading.cpp#42). It actually only uses these samples (and doesn't interpolate between them) apparently in order to make hard stops look hard, even if they're in the wrong place. I didn't realize it did this and explains why sometimes it looks strange. (It only does this for axial and radial, the others are evaluated per-pixel.)

Poppler is interesting because it depends on the OutputDev implementation, of which there are several of varying quality. I think the one I'm used to looking at [in Okular](https://gitlab.freedesktop.org/poppler/poppler/-/blob/96067bdbc2307decacc8d526795fd2dacf617f5d/qt5/src/QPainterOutputDev.cc#L709) seems to create a gradient of at most 256 samples but chooses them through bisection, which seems like it could go very wrong (differently, but like my initial approach at this MR) if the midpoint accidentally happens to match when it really doesn't. Interestingly, [the base implementation which this overrides](https://gitlab.freedesktop.org/poppler/poppler/-/blob/96067bdbc2307decacc8d526795fd2dacf617f5d/poppler/Gfx.cc#L2700) has a note about this and tries to work around this by sampling two points while bisecting. I would call this approach "optimistic" in the sense that it tries to evaluate the function fewer times trying to confirm its guesses, where as the current approach of this MR is "pessimistic" in that it just picks a sampling frequency, does the samples, and compresses the function. The "optimistic" approach does have the advantage of potentially fewer samples of the function (in return for being more easily tricked) and possibly more adaptive (more samples in higher frequency areas, like narrowing down hard stops).

In any event, after looking at the numbers it doesn't look like evaluating the function is any real bottleneck in practice.

Reply:
Getting back around to profiling, using a [patch to Skia](https://skia-review.googlesource.com/c/skia/+/658578) in a local build of Chromium, I was able to create a [PDF](https://github.com/mozilla/pdf.js/files/11023153/Noto.Color.Emoji.-.Google.Fonts.-.Chrome.112.-.Full.-.PSFuncs.pdf) of Noto Color Emoji COLRv1 with Type 4 (PostScript) Functions for everything. Basically issue16127.pdf but with all Type 4 Functions instead of Type 3 and Type 2.

When profiling one can see significant time spent in the DOM worker thread. With this change a given sampling shows that most of the time is spent in
```
27,717 Worker
  20,244 RadialAxialShading
    16,753 constructPostScriptFn
      10,865 execute
  2,260 getRgb
```

without this change it looks like

```
18,692 Worker
  921 RadialAxialShading
    277 constructPostScriptFn
      195 execute
  46 getRgb
```

which appears to be about what one would expect from 82x. Taking some more profiles of issue16127.pdf it appears there is a similar change in the DOM worker thread, but the Type 2 and 3 functions were already so fast that it is much less noticeable.

All that to say that, yes, this does take longer (as expected, it is doing more samples). It is, of course, possible to tweak the number of samples.

As a note if it is known that the function is made up of Type 0, 2, or 3 Functions, it is fairly simple to determine where to sample. In theory this is somewhat possible with many Type 4 functions (without `sin`, `cos`, and maybe one or two others) since one should be able to express the first and second derivatives of the functions (multiplying out by possible `if` and `ifelse`) as polynomials and root solve them. This is something of a longer term research project than I'm willing to sign up for at the moment though.

It may be faster to attempt the "optimistic" approach of testing and subdividing, but the tests would need to at least be kinda pseudo random (at least not periodic).